File: rust/src/api/metadata.rs
use flutter_rust_bridge::frb;
use mc_launcher_core::loader;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct VanillaVersion {
    pub id: String,
    #[serde(rename = "type")]
    pub version_type: String,
}

#[derive(Debug, Deserialize)]
struct VanillaManifest {
    versions: Vec<VanillaVersion>,
}

pub fn get_vanilla_versions() -> anyhow::Result<Vec<VanillaVersion>> {
    let manifest: VanillaManifest = reqwest::blocking::get("https://launchermeta.mojang.com/mc/game/version_manifest_v2.json")?.json()?;
    Ok(manifest.versions)
}

pub fn get_fabric_loaders() -> anyhow::Result<Vec<String>> {
    let loaders = loader::fabric::list_loader_versions()?;
    Ok(loaders.into_iter().map(|l| l.version).collect())
}

pub fn get_quilt_loaders() -> anyhow::Result<Vec<String>> {
    let loaders = loader::quilt::list_loader_versions()?;
    Ok(loaders.into_iter().map(|l| l.version).collect())
}

pub fn get_forge_versions() -> anyhow::Result<Vec<String>> {
    loader::forge::list_forge_versions().map_err(Into::into)
}

pub fn get_neoforge_versions() -> anyhow::Result<Vec<String>> {
    loader::neoforge::list_neoforge_versions().map_err(Into::into)
}

File: lib/src/rust/api/metadata.dart


import '../frb_generated.dart';
import 'package:flutter_rust_bridge/flutter_rust_bridge_for_generated.dart';

Future<List<VanillaVersion>> getVanillaVersions() =>
    RustLib.instance.api.crateApiMetadataGetVanillaVersions();

Future<List<String>> getFabricLoaders() =>
    RustLib.instance.api.crateApiMetadataGetFabricLoaders();

Future<List<String>> getQuiltLoaders() =>
    RustLib.instance.api.crateApiMetadataGetQuiltLoaders();

Future<List<String>> getForgeVersions() =>
    RustLib.instance.api.crateApiMetadataGetForgeVersions();

Future<List<String>> getNeoforgeVersions() =>
    RustLib.instance.api.crateApiMetadataGetNeoforgeVersions();

class VanillaVersion {
  final String id;
  final String versionType;

  const VanillaVersion({
    required this.id,
    required this.versionType,
  });

  @override
  int get hashCode => id.hashCode ^ versionType.hashCode;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is VanillaVersion &&
          runtimeType == other.runtimeType &&
          id == other.id &&
          versionType == other.versionType;
}
